### PR TITLE
Fix simrank_similarity with directed graph input

### PR DESCRIPTION
**Pull request by ryan-duve**
User @Carghaez pointed out the SimRank function is not behaving properly in [Issue 3953](https://github.com/networkx/networkx/issues/3953).  It appears the algorithm was trying to iterate over out-neighbor pairs.  On the [top of page 540 of the original paper](https://dl.acm.org/doi/pdf/10.1145/775047.775126) it says the algorithm should iterate over in-neighbor pairs.

This fix is performed by changing `G[<node>]` to `G.predecessors(<node>)`.

**Note to maintainers:** I had to change several tests after the fix.  A cycle graph using `DiGraph` works, but it gives trivial results since no nodes have common nodes pointing to them.  For a non-trivial choice, I chose to go with the implementation in the paper (the same one @Carghaez used in the issue he raised).  The `expected` strings match results from both the numbers in the paper and the `simrank_similarity_numpy` function.

**Disclaimer**: I just learned about SimRank today and am by no means an expert, so please review this change with a critical eye.

### CODE REVIEW

**Comment by Carghaez** [comment]
I also found this bug a few days ago and I confirm that is the right fix!

**Comment by dschult** [comment]
This looks like you changes from using ```G[u]``` to ```G,predecessors(u)``` but that won't work for undirected graphs. Then you removed the tests of undirected graphs and changed them to directed graphs. So, while it is passing all the tests now, it still won't work for undirected graphs. Am I correct, or am I missing an if-clause somewhere that makes this only get used for directed graphs?  I'll get a chance to dig around in this sometime soon. But it would reassure me if you left the current tests of cycle_graphs and just added the new tests of DiGraphs.  It's Ok to have two assert statements in the same test. Just leave the existing one and add the new one.

Let me know if I'm missing something.

**Comment by Carghaez** [comment]
@dschult I already fix that issue here https://github.com/networkx/networkx/pull/3962

**Comment by ryan-duve** [comment]
@dschult Here's what happened:

1. I opened the original PR supporting DiGraphs but breaking tests for undirected graphs because I mistakenly thought undirected graphs weren't supported by SimRank.
1. @Carghaez recognized the error and created a regression PR with a conditional to support both directed and undirected graphs.
1. This morning I pulled in the changes from that PR into this one.
1. This morning I restored the tests for undirected graphs and documentation, putting several assert in the same tests like you suggested.

The tests for the PR are now additive only and directed graphs are supported.

File: networkx/algorithms/similarity.py [hunk 1]
```diff
@@ -1270,7 +1270,11 @@ def avg_sim(s):
         return sum(newsim[w][x] for (w, x) in s) / len(s) if s else 0.0
 
     def sim(u, v):
-        return importance_factor * avg_sim(list(product(G[u], G[v])))
+        if nx.classes.function.is_directed(G):
```
Comment (dschult) [inline networkx/algorithms/similarity.py:1273]: Maybe combine avg_sim and sim using something like:

```python
Gadj = G.pred if G.is_directed() else G.adj
def sim(u, v):
    pairs = list(product(Gadj[u], Gadj[v]))
    if not pairs:
        return 0.0
    return importance_factor * sum(newsim[w][x] for w, x in pairs) / len(pairs)
```

Comment (ryan-duve) [reply networkx/algorithms/similarity.py]: @dschult Cool to learn about `G.pred` and `G.adj`.  I change the conditional per your recommendation and pushed the change.

At first I also combined the functions, but as a new user of NetworkX my opinion is this made the SimRank function harder to read and understand.  I left them as separate for now so you can review.  If it is more in line with NetworkX style to combine the functions, let me know and I will make that change, too.

**Comment by Carghaez** [comment]
Looks good for me! +1

**Pull request merged**